### PR TITLE
Feat/django session replace cookies

### PR DESCRIPTION
**Pull request by muhammad-ali-e**
## What

- This PR enhances Django session functionality
- Removal of Unnecessary Cookie Creations: Unwanted cookie creations have been removed from the session handling process.
- Cleanup of Callback Functions: Callback functions have been cleaned up for improved code readability and maintainability.
- Refactoring: few functions have been refactored to enhance code efficiency and clarity.
- SESSION_ENGINE: changed to cached_db. [First preffer from cache then DB]

## Why

- The primary goal of these enhancements is to streamline the session management process within Django, thereby reducing complexity and avoiding the proliferation of multiple session types and cookies.

## How

- To achieve this goal, we have opted to leverage Django's built-in session management features for authentication and session handling, ensuring a standardized and consistent approach across the application.

## Notes on Testing

- Please conduct thorough testing covering all aspects of session management and authentication to ensure the changes are functioning as expected.

## Related Issues or PRs
- **Cloud 56**
- https://github.com/Zipstack/unstract/pull/264

## Checklist

I have read and understood the [Contribution Guidelines]().

### CODE REVIEW

File: backend/account/custom_auth_middleware.py [hunk 1]
```diff
@@ -38,73 +26,19 @@ def __call__(self, request: HttpRequest) -> HttpResponse:
...
+
+        is_authenticated = auth_service.is_authenticated(request)
+
+        if is_authenticated:
             StateStore.set(Common.LOG_EVENTS_ID, request.session.session_key)
             response = self.get_response(request)
             StateStore.clear(Common.LOG_EVENTS_ID)
 
             return response
         return JsonResponse({"message": "Unauthorized"}, status=401)
```
Comment (chandrasekharan-zipstack) [inline backend/account/custom_auth_middleware.py:44]: Do you remember why this returns a `JsonResponse` instead of a `HttpResponse`?

File: backend/account/constants.py [hunk 2]
```diff
@@ -67,8 +67,15 @@ class PluginConfig:
 
 
 class AuthorizationErrorCode:
-    """Error code reference
-    frontend/src/components/error/GenericError/GenericError.jsx."""
+    """Error codes
```
Comment (chandrasekharan-zipstack) [inline backend/account/constants.py:70]: @muhammad-ali-e shouldn't these be part of the plugin ideally?
CC: @hari-kuriakose

File: backend/account/serializer.py [hunk 3]
```diff
@@ -109,3 +109,10 @@ def validate_password(self, value: Optional[str]) -> str:
                 "Password must be at least 3 characters long."
             )
         return value
+
+
+class UserSessionResponseSerializer(serializers.Serializer):
+    id = serializers.CharField()
```
Comment (chandrasekharan-zipstack) [inline backend/account/serializer.py:115]: @muhammad-ali-e shouldn't this be a `UUIDField()`?

File: backend/account/views.py [hunk 4]
```diff
@@ -97,6 +103,38 @@ def set_organization(request: Request, id: str) -> Response:
...
+@api_view(["GET"])
+def get_session_data(request: Request) -> Response:
+    try:
+        response = make_session_response(request)
+        return Response(
+            status=status.HTTP_201_CREATED,
+            data=response,
+        )
+
+    except Exception as error:
```
Comment (chandrasekharan-zipstack) [inline backend/account/views.py:115]: Please try to catch specific exceptions and handle them accordingly - uncaught exceptions would anyway be logged with a traceback and returned as 500 errors by django. But we'd have more information regarding the error itself in our logs. We'll have to move out of this practice as we refactor / improve our error handling

Comment (chandrasekharan-zipstack) [reply backend/account/views.py:115]: In this case, getting rid of the `try-catch` block would make no difference in behaviour and in-fact help improve logging with a traceback

File: backend/backend/settings/base.py [hunk 5]
```diff
@@ -404,11 +406,6 @@ def get_required_setting(
 # White lists workflow-api-deployment path
 WHITELISTED_PATHS.append(f"/{API_DEPLOYMENT_PATH_PREFIX}")
 
-# White list paths under tenant paths
```
Comment (chandrasekharan-zipstack) [inline backend/backend/settings/base.py:407]: @muhammad-ali-e why was this removed? Has it been moved elsewhere?

File: backend/file_management/views.py [hunk 6]
```diff
@@ -113,7 +114,7 @@ def delete(self, request: HttpRequest) -> Response:
         file_name: str = document.document_name
         tool_id: str = serializer.validated_data.get("tool_id")
         file_path = FileManagerHelper.handle_sub_directory_for_tenants(
-            request.org_id,
+            UserSessionUtils.get_organization_id(request),
```
Comment (chandrasekharan-zipstack) [inline backend/file_management/views.py:117]: @muhammad-ali-e what's the difference between `request.org_id` and `UserSessionUtils.get_organization_id(request)`?
Can we set `request.org_id` in the middleware itself and continue to use the same?

File: backend/account/authentication_helper.py [hunk 7]
```diff
@@ -25,6 +27,16 @@ def list_of_members_from_user_model(
 
         return members
 
+    @staticmethod
+    def get_or_create_user_by_email(
+        user_id: str, email: str
+    ) -> Union[User, OrganizationMember]:
```
[the remark this answers is not in the document]
Comment (muhammad-ali-e) [reply backend/account/authentication_helper.py]: It was there in another file. I think previously we used this function in multiple contexts (user and organization). But I think now we need User only, So changed this to User

File: backend/account/custom_auth_middleware.py (hunk 1 above)
Comment (muhammad-ali-e) [reply backend/account/custom_auth_middleware.py:44]: I think JsonResponse is is mandatory in the case of json data. view sets and api_view decorators are handled to send JsonResponse.

https://docs.djangoproject.com/en/5.0/ref/request-response/#jsonresponse-objects
https://docs.datadoghq.com/code_analysis/static_analysis_rules/python-django/http-response-with-json-dumps/

File: backend/account/constants.py (hunk 2 above)
Comment (muhammad-ali-e) [reply backend/account/constants.py:70]: This PR doesn't have much changes here except here added the docstring.
I feels like both ways,
 Is this part of a plugin? It seems to be utilized exclusively for multi-tenancy (currently enterprise-only).
 Is this part of OSS? It incorporates certain UI components and could potentially be reused with other plugins and OSS.
 @johnyrahul

File: backend/account/serializer.py (hunk 3 above)
Comment (muhammad-ali-e) [reply backend/account/serializer.py]: Actually we need IntegerFeild here. Changed

File: backend/account/views.py (hunk 4 above)
Comment (muhammad-ali-e) [reply backend/account/views.py]: Removed genric excedption catch from here.

File: backend/backend/settings/base.py (hunk 5 above)
Comment (muhammad-ali-e) [reply backend/backend/settings/base.py:407]: Tested with some scenarios , Here we don't need these list because of django session authentication handled this. Please let me know are there any case that break this changes. @gopalsamyperumal @VikashPratheepan

File: backend/file_management/views.py (hunk 6 above)
Comment (muhammad-ali-e) [reply backend/file_management/views.py:117]: We've transitioned to utilizing Django session instead of storing data in the request model. Session data persists across requests, ensuring consistency. This eliminates the need to repeatedly store data in the request model with each call.

File: backend/backend/settings/base.py (hunk 5 above)
Comment (chandrasekharan-zipstack) [reply backend/backend/settings/base.py:407]: @muhammad-ali-e specifically did you check for these 2 scenarios?
1. GDrive OAuth
1. Accessing documentation with `http://localhost:8000/api/v1/doc/`

File: backend/file_management/views.py (hunk 6 above)
Comment (muhammad-ali-e) [reply backend/file_management/views.py:117]: There we are not using request.org.  they are public urls .

**Review by chandrasekharan-zipstack — APPROVED** [review]

File: frontend/src/hooks/useSessionValid.js [hunk 8]
```diff
@@ -35,16 +33,13 @@ function useSessionValid() {
       if (!orgs?.length) {
         throw Error("Organizations not available.");
       }
-      if (orgs?.length > 1 && !signedInOrgId?.length) {
+      if (orgs?.length > 1 && signedInOrgId && signedInOrgId === "public") {
```
Comment (tahierhussain) [inline frontend/src/hooks/useSessionValid.js:36]: @muhammad-ali-e Can we simplify this condition to this,

`orgs?.length > 1 && signedInOrgId === "public"`

File: frontend/src/helpers/GetUserSession.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+      method: "GET",
+      url: "/api/v1/session",
+    };
+    console.log("requestOptions", requestOptions);
+    const res = await axios(requestOptions);
+    console.log("res", res);
+    return res.data;
+  } catch (error) {
+    const handleException = useExceptionHandler();
+    handleException(error, "Error while getting session");
```
Comment (tahierhussain) [inline frontend/src/helpers/GetUserSession.js:16]: @muhammad-ali-e Can you please refer [this code](https://github.com/Zipstack/unstract/blob/ad0ffcdb642ee94d0870f1eb3964338cd75624fc/frontend/src/components/custom-tools/tool-ide/ToolIde.jsx#L106) for handling exception?

File: frontend/src/helpers/GetCookie.js [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    for (const element of cookies) {
+      const cookie = element.trim();
+      if (cookie.substring(0, name.length + 1) === name + "=") {
+        cookieValue = decodeURIComponent(cookie.substring(name.length + 1));
+        break;
+      }
+    }
+  }
+  return cookieValue;
+}
```
Comment (tahierhussain) [inline frontend/src/helpers/GetCookie.js:14]: @muhammad-ali-e Suggestion: We can avoid nested if conditions by modifying the code like below,

```
function getCookie(name) {
    let cookieValue = null;

    if (!document.cookie) {
      return cookieValue;
    }

    const cookies = document.cookie.split(";");
    for (const element of cookies) {
      const cookie = element.trim();
      if (cookie.substring(0, name.length + 1) === name + "=") {
        cookieValue = decodeURIComponent(cookie.substring(name.length + 1));
        break;
      }
    }
    return cookieValue;
  }
```

**Review by tahierhussain — CHANGES_REQUESTED** [review]

File: frontend/src/components/set-org/SetOrg.jsx [hunk 11]
```diff
@@ -23,23 +25,29 @@ function SetOrg() {
...
-      navigate("/");
-    }
-  }, []);
+    const fetchData = async () => {
+      try {
+        const userSessionData = await userSession();
+        const signedInOrgId = userSessionData?.organization_id;
+        if (state === null || (signedInOrgId && signedInOrgId !== "public")) {
+          navigate("/");
+          return;
```
Comment (tahierhussain) [inline frontend/src/components/set-org/SetOrg.jsx:34]: @muhammad-ali-e I think it should be safe to remove the `return` statement, as the logic concludes at this point.

File: frontend/src/components/set-org/SetOrg.jsx [hunk 12]
```diff
@@ -23,23 +25,29 @@ function SetOrg() {
...
+        }
+      } catch (error) {
+        navigate("/");
+        return;
```
Comment (tahierhussain) [inline frontend/src/components/set-org/SetOrg.jsx:38]: @muhammad-ali-e Same here

File: frontend/src/components/set-org/SetOrg.jsx [hunk 13]
```diff
@@ -23,23 +25,29 @@ function SetOrg() {
   const handleException = useExceptionHandler();
 
   useEffect(() => {
-    if (state === null || signedInOrgId) {
-      navigate("/");
-    }
-  }, []);
+    const fetchData = async () => {
```
Comment (tahierhussain) [inline frontend/src/components/set-org/SetOrg.jsx:28]: @muhammad-ali-e Can we declare this function outside the `useEffect` hook?

Reason: The function will be recreated each time the `useEffect` hook runs. In our case, each time the `state` or `navigate` value changes, this `useEffect` hook will get triggered, and the `fetchData` function will be recreated.

**Review by tahierhussain — CHANGES_REQUESTED** [review]

File: frontend/src/components/set-org/SetOrg.jsx (hunk 13 above)
Comment (muhammad-ali-e) [reply frontend/src/components/set-org/SetOrg.jsx:28]: @tahierhussain I tried But it have to be inside useEffect
https://react.dev/reference/react/useEffect#fetching-data-with-effects:~:text=a%20cleanup%20function%3A-,App.js,-Reset

File: frontend/src/components/set-org/SetOrg.jsx (hunk 12 above)
Comment (muhammad-ali-e) [reply frontend/src/components/set-org/SetOrg.jsx]: updated

File: frontend/src/components/set-org/SetOrg.jsx (hunk 11 above)
Comment (muhammad-ali-e) [reply frontend/src/components/set-org/SetOrg.jsx]: updated

File: frontend/src/helpers/GetCookie.js (hunk 10 above)
Comment (muhammad-ali-e) [reply frontend/src/helpers/GetCookie.js]: Replaced this logic with package  https://www.npmjs.com/package/js-cookie

File: frontend/src/helpers/GetUserSession.js (hunk 9 above)
Comment (muhammad-ali-e) [reply frontend/src/helpers/GetUserSession.js]: updated..  But I have few doubts here. We can clarify later

File: frontend/src/hooks/useSessionValid.js (hunk 8 above)
Comment (muhammad-ali-e) [reply frontend/src/hooks/useSessionValid.js]: updated here. I think It's wrong to check public here.

File: frontend/src/helpers/GetUserSession.js [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+const userSession = async () => {
+  try {
+    const requestOptions = {
+      method: "GET",
+      url: "/api/v1/session",
+    };
+    console.log("requestOptions", requestOptions);
+    const res = await axios(requestOptions);
+    console.log("res", res);
```
[the remark this answers is not in the document]
Comment (muhammad-ali-e) [reply frontend/src/helpers/GetUserSession.js]: removed

File: frontend/src/helpers/GetUserSession.js [hunk 15]
```diff
@@ -0,0 +1,20 @@
+import axios from "axios";
+import { useExceptionHandler } from "../hooks/useExceptionHandler.jsx";
+
+const userSession = async () => {
+  try {
+    const requestOptions = {
+      method: "GET",
+      url: "/api/v1/session",
+    };
+    console.log("requestOptions", requestOptions);
```
[the remark this answers is not in the document]
Comment (muhammad-ali-e) [reply frontend/src/helpers/GetUserSession.js]: removed

File: frontend/src/components/set-org/SetOrg.jsx (hunk 13 above)
Comment (tahierhussain) [reply frontend/src/components/set-org/SetOrg.jsx:28]: @muhammad-ali-e Let me look into this and we can address this improvement in a separate PR. I'm resolving this conversation and approving this PR so that it's not blocked.

cc: @nehabagdia

**Review by tahierhussain — APPROVED** [review]
FE changes LGTM
